### PR TITLE
Rework DataArray internals

### LINKED ISSUE #367: ds['time.time'] is broken
As noted in https://github.com/xray/xray/issues/364:
```
In [32]: t = pd.date_range('2000-01-01', periods=10, freq='H')

In [33]: time = xray.DataArray(t, name='time', dims='time')

In [34]: time['time.time']
Out[34]:
<xray.DataArray 'time' (time: 10)>
array(['1999-12-31T16:00:00.000000000-0800',
       '1999-12-31T17:00:00.000000000-0800',
       '1999-12-31T18:00:00.000000000-0800',
       '1999-12-31T19:00:00.000000000-0800',
       '1999-12-31T20:00:00.000000000-0800',
       '1999-12-31T21:00:00.000000000-0800',
       '1999-12-31T22:00:00.000000000-0800',
       '1999-12-31T23:00:00.000000000-0800',
       '2000-01-01T00:00:00.000000000-0800',
       '2000-01-01T01:00:00.000000000-0800'], dtype='datetime64[ns]')
Coordinates:
  * time     (time) datetime64[ns] 2000-01-01 2000-01-01T01:00:00 ...
```

It should return an array of `datetime.time` objects.

### LINKED ISSUE #634: Unexpected behavior by diff when applied to coordinate DataArray
`DataArray.diff`, when applied to a coordinate, just returns the coordinate itself with one index truncated.  Consider the following:
```
In [5]: arr = xray.DataArray(range(0, 20, 2), dims=['lon'], coords=[range(10)])
In [6]: arr.diff('lon')
<xray.DataArray (lon: 9)>
array([2, 2, 2, 2, 2, 2, 2, 2, 2])
Coordinates:
  * lon      (lon) int64 1 2 3 4 5 6 7 8 9

In [7]: arr['lon'].diff('lon')
<xray.DataArray 'lon' (lon: 9)>
array([1, 2, 3, 4, 5, 6, 7, 8, 9])
Coordinates:
  * lon      (lon) int64 1 2 3 4 5 6 7 8 9
```

Is this the intended behavior?  The documentation doesn't mention anything about this, and its counter-intuitive, so I'm wondering if its a bug instead.  Even if it is intended, I personally would like to be able to use a coordinate array's diff on itself to get its spacing, e.g. for use as the denominator in finite differencing approximations to derivatives.

Thanks!

### LINKED ISSUE #649: error when using broadcast_arrays with coordinates
I frequently use ``broadcast_arrays`` to to feed xray variables to non-xray libraries (e.g. [gsw](https://github.com/TEOS-10/python-gsw).) Often I need to broadcast the coordinates and variables in order to do call functions that take both as arguments.

I have found that ``broadcast_arrays`` doesn't work as I expect with coordinates. For example

```python
import xray
import numpy as np
ds = xray.Dataset({'a': (['y','x'], np.ones((20,10)))},
                  coords={'x': (['x'], np.arange(10)),
                          'y': (['y'], np.arange(20))})
xbc, ybc, abc = xray.broadcast_arrays(ds.x, ds.y, ds.a)
```

This raises ``ValueError: an index variable must be defined with 1-dimensional data``. 

If I change the last line to
```python
xbc, ybc, abc = xray.broadcast_arrays(1*ds.x, 1*ds.y, ds.a)
```
it works fine.

### PR DESCRIPTION
**Pull request by shoyer**
Fixes #367
Fixes #634
Fixes #649

The internal data model used by `DataArray` has been rewritten to fix several outstanding issues (#367, #634 and [this stackoverflow report](http://stackoverflow.com/questions/33158558/python-xray-extract-first-and-last-time-value-within-each-month-of-a-timeseries)). Namely, if a DataArray has the same name as one of its coordinates, the array and the coordinate no longer share the same data.

This means that creating a DataArray with the same ``name`` as one of its dimensions no longer automatically uses that array to label the corresponding coordinate. You will now need to provide coordinate labels explicitly. Here's the old behavior:

    In [2]: xray.DataArray([4, 5, 6], dims='x', name='x')
    Out[2]:
    <xray.DataArray 'x' (x: 3)>
    array([4, 5, 6])
    Coordinates:
      * x        (x) int64 4 5 6

and the new behavior (compare the values of the ``x`` coordinate):

    In [2]: xray.DataArray([4, 5, 6], dims='x', name='x')
    Out[2]:
    <xray.DataArray 'x' (x: 3)>
    array([4, 5, 6])
    Coordinates:
      * x        (x) int64 0 1 2

It's also no longer possible to convert a DataArray to a Dataset with `DataArray.to_dataset` if it is unnamed. This will now raise ``ValueError``. If the array is unnamed, you need to supply the ``name`` argument.

### CODE REVIEW

**Comment by jhamman** [comment]
@shoyer - I read this mainly trying to get a better idea of the internal DataArray data model.  The code itself looks great.  My main two comments on the refactor are:

1.  The use of the `_ThisArray` singleton seems a bit clunky to me.  This is my first time interacting with this approach so maybe it is the best way, but I don't know much about it.
2.  This all seems to be complicated by the DataArrays's use of the Dataset internally to store coordinates.  Not for this PR, but I wonder if there is a way to clear up the lines between the two objects.

All in all, impressive work.

**Comment by shoyer** [comment]
Indeed, I wonder if it would make sense to decouple DataArray from Dataset by storing the state on two (protected) attributes:
- `variable`: the variable for this array
- `coords`: an ordered dict of coordinates

The main downside is that we add a bit more redundant code (e.g., to loop over all variables in `.sel`). But on the plus side our data model is much more similar to the public API, which would probably make things easier to understand, especially for new contributors -- they don't have to learn Dataset to learn DataArray.

**Comment by maximilianr** [comment]
>Indeed, I wonder if it would make sense to decouple DataArray from Dataset by storing the state on two (protected) attributes:

As a newbie, :+1:. I took some time to figure out why a `DataArray` _contained_ a `Dataset`. 

>The main downside is that we add a bit more redundant code (e.g., to loop over all variables in .sel).

Low confidence, but you could have a common ancestor (`XRayContainer`, etc) which contained the similar code or a `._stuff` attribute that's iterated over

File: xray/core/dataarray.py [hunk 1]
```diff
@@ -233,11 +242,13 @@ def subset(dim, label):
         del coords[dim]
         return Dataset(variables, coords, self.attrs)
 
-    def _to_dataset_whole(self, name):
+    def _to_dataset_whole(self, name=None):
         if name is None:
```
Comment (maximilianr) [inline xray/core/dataarray.py]: Very minor, but `name = name or self.name` might be clearer / more pythonic than the repeated `if name is None:`

**Comment by jhamman** [comment]
@shoyer - do you have a feel for how difficult it would be to go the `variable`/`coords` route?

**Comment by shoyer** [comment]
> @shoyer - do you have a feel for how difficult it would be to go the variable/coords route?

Hmm. Might not be so bad now that I've already gone through the trouble of thinking what these new tests should look like. I'll give it a shot tonight and see how it goes...

**Comment by shoyer** [comment]
I realize now that changing the internal representation for DataArray doesn't mean we need to rewrite how every routine works. We can still convert dataarrays to a dataset when convenient -- it just means we'll need to use a method to do so instead of modifying `._dataset`. For example:
```diff
@@ -490,8 +522,8 @@ class DataArray(AbstractArray, BaseDataObject):
         dataset. Otherwise, a shallow copy is made, so each variable in the new
         array's dataset is also a variable in this array's dataset.
         """
-        ds = self._dataset.copy(deep=deep)
-        return self._with_replaced_dataset(ds)
+        ds = self._to_temp_dataset().copy(deep=deep)
+        return self._new_from_temp_dataset(ds)

     def __copy__(self):
         return self.copy(deep=False)
```

**Comment by shoyer** [comment]
OK, latest commit changes DataArray's internals to rely on `_variable` and `_coords` instead of `_dataset`. There's still a bit more `_to_temp_dataset()` than ideal, but hopefully the data model makes more sense now...

**Comment by shoyer** [comment]
This is ready for review if anyone wants to take another look.

File: xray/core/dataarray.py [hunk 2]
```diff
@@ -913,9 +955,19 @@ def to_dataframe(self):
...
-        # TODO: add a 'name' parameter
+        if name is None:
+            name = self.name
+        if name is None:
+            raise ValueError('cannot convert an unnamed DataArray to a '
+                             'DataFrame: use the ``name`` parameter')
+
         dims = OrderedDict(zip(self.dims, self.shape))
-        return self._dataset._to_dataframe(dims)
+        unique_name = '__unique_name_identifier_z98xfz98xugfg73ho__'
```
Comment (jhamman) [inline xray/core/dataarray.py]: what's going on here?

Comment (shoyer) [inline xray/core/dataarray.py]: Good catch -- needed explanation. Let me know if the comments I added help.

**Comment by jhamman** [comment]
@shoyer - I don't have any more inline comments.  There is one failing test and there are merge conflicts, once those are addressed, I'll take one more brief look.

**Comment by shoyer** [comment]
Rebased and tests are passing.

**Comment by jhamman** [comment]
lgtm, go ahead and merge.

**Comment by MaximilianR** [comment]
:clap: 
I'll try and get my PR in this weekend for the pandas wrapping

**Pull request merged**